arbnode/message_pruner.go

```diff
@@ -25,107 +28,206 @@ type MessagePruner struct {
+// returns number of batch for which report was found (meaning - it should not be pruned)
+// if not found - returns maxUint64 (no limit on pruning)
+func (m *MessagePruner) findBatchReport(ctx context.Context, delayedMsgStart uint64) (uint64, error) {
+	searchLimit := m.config().SearchBatchReportLimit
+	if searchLimit < 0 {
+		return math.MaxUint64, nil
+	}
+	delayedCount, err := m.inboxTracker.GetDelayedCount()
 	if err != nil {
-		log.Error("error getting latest confirmed node info", "node", latestConfirmedNode, "err", err)
-		return m.config().MessagePruneInterval
+		return 0, err
+	}
+	if delayedCount <= delayedMsgStart {
+		return 0, errors.New("delayedCount behind pruning target")
+	}
+	searchUpTil := delayedCount
+	searchUpLimit := delayedMsgStart + uint64(searchLimit)
+	if searchLimit > 0 && searchUpLimit < searchUpTil {
+		searchUpTil = searchUpLimit
 	}
-	endBatchCount := nodeInfo.Assertion.AfterState.GlobalState.Batch
-	if endBatchCount == 0 {
-		return m.config().MessagePruneInterval
+	for delayed := delayedMsgStart; delayed < searchUpTil; delayed++ {
+		if ctx.Err() != nil {
+			return 0, ctx.Err()
+		}
+		msg, err := m.inboxTracker.GetDelayedMessage(delayed)
+		if err != nil {
+			return 0, err
+		}
+		if msg.Header.Kind == arbostypes.L1MessageType_BatchPostingReport {
+			_, _, _, batchNum, _, _, err := arbostypes.ParseBatchPostingReportMessageFields(bytes.NewReader(msg.L2msg))
+			if err != nil {
+				return 0, fmt.Errorf("trying to parse batch-posting report: %w", err)
+			}
+			return batchNum, nil
+		}
 	}
-	endBatchMetadata, err := m.inboxTracker.GetBatchMetadata(endBatchCount - 1)
+	searchDownLimit := uint64(0)
+	if searchLimit > 0 {
+		searchedUp := searchUpTil - delayedMsgStart
+		limitRemaining := uint64(searchLimit) - searchedUp
+		if limitRemaining < delayedMsgStart {
+			searchDownLimit = delayedMsgStart - limitRemaining
+		}
+	}
+	for delayed := delayedMsgStart - 1; delayed >= searchDownLimit; delayed-- {
+		if ctx.Err() != nil {
+			return 0, ctx.Err()
+		}
+		msg, err := m.inboxTracker.GetDelayedMessage(delayed)
+		if errors.Is(err, AccumulatorNotFoundErr) {
+			// older delayed probably pruned - assume we won't find a report
+			return math.MaxUint64, nil
+		}
+		if err != nil {
+			return 0, err
+		}
+		if msg.Header.Kind == arbostypes.L1MessageType_BatchPostingReport {
+			_, _, _, batchNum, _, _, err := arbostypes.ParseBatchPostingReportMessageFields(bytes.NewReader(msg.L2msg))
+			if err != nil {
+				return 0, fmt.Errorf("trying to parse batch-posting report: %w", err)
+			}
+			// found below delayedMessage - so batchnum can be pruned but above it cannot
+			return batchNum + 1, nil
+		}
+	}
+	return math.MaxUint64, nil
+}
+
+func (m *MessagePruner) prune(ctx context.Context, count arbutil.MessageIndex, globalState validator.GoGlobalState) error {
+	trimBatchCount := globalState.Batch
+	minBatchesLeft := m.config().MinBatchesLeft
+	if trimBatchCount < minBatchesLeft {
```

Comment:
Theoretically I think this comparison should be against `batchCount` not `trimBatchCount`, but this works too in practice